atomics/include/desul/atomics/SYCL.hpp

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright (c) 2019, Lawrence Livermore National Security, LLC
+and DESUL project contributors. See the COPYRIGHT file for details.
+Source: https://github.com/desul/desul
+
+SPDX-License-Identifier: (BSD-3-Clause)
+*/
+#ifndef DESUL_ATOMICS_SYCL_HPP_
+#define DESUL_ATOMICS_SYCL_HPP_
+
+#ifdef DESUL_HAVE_SYCL_ATOMICS
+#include "desul/atomics/Common.hpp"
+
+namespace desul {
+namespace Impl {
+template<class T>
+struct is_sycl_atomic_type {
+  static constexpr bool value = std::is_same<T, int>::value ||
+                                std::is_same<T, unsigned int>::value ||
+				std::is_same<T, long>::value ||
+				std::is_same<T, unsigned long>::value ||
+				std::is_same<T, long long>::value ||
+                                std::is_same<T, unsigned long long int>::value ||
+				std::is_same<T, float>::value ||
+				std::is_same<T, double>::value;
+};
+} // Impl
+
+// Atomic Add
+template<class T, class MemoryOrder/*, class MemoryScope*/>
+inline
+typename std::enable_if<Impl::is_sycl_atomic_type<T>::value,T>::type
+atomic_fetch_add(T* dest, T val, MemoryOrder, MemoryScopeDevice) {
+  DESUL_SYCL_NAMESPACE::atomic_ref<
+    T,
+    DesulToSYCLMemoryOrder<MemoryOrder>::value,
+    DesulToSYCLMemoryScope<MemoryScopeDevice>::value,
+    sycl::access::address_space::global_device_space>
+  dest_ref(*dest);
+  return dest_ref.fetch_add(val);
+}
+
+// Atomic Sub
+template<class T, class MemoryOrder/*, class MemoryScope*/>
+inline
+typename std::enable_if<Impl::is_sycl_atomic_type<T>::value,T>::type
+atomic_fetch_sub(T* dest, T val, MemoryOrder, MemoryScopeDevice) {
+  DESUL_SYCL_NAMESPACE::atomic_ref<
+    T,
+    DesulToSYCLMemoryOrder<MemoryOrder>::value,
+    DesulToSYCLMemoryScope<MemoryScopeDevice>::value,
+    sycl::access::address_space::global_device_space>
+  dest_ref(*dest);
+  return dest_ref.fetch_sub(val);
+}
+
+// Atomic Inc
+template<class MemoryOrder/*, class MemoryScope*/>
+inline
+unsigned int atomic_fetch_inc(unsigned int* dest, unsigned int val, MemoryOrder memory_order, MemoryScopeDevice memory_scope) {
+  return atomic_fetch_add(dest, val, memory_order, memory_scope);
+}
```

Comment:
Why is it not templated?  Also, I see that CUDA and HIP do that as well, but the signature makes no sense.  `val` is always `1` or we should be using `atomic_fetch_add`.

Reply:
This function has a different signature for the generic implementation that doesn't take a value (and calls `atomic_fetch_add` with `value=1`). For `CUDA`, `atomicInc` is the same as `atomicAdd` but only works on 32-bit objects.
IMHO, we should just delete this one and the corresponding ones for `CUDA` and `HIP`.

Opinions on this?

Reply:
I actually just got someone working on fixing the generic version of that yesterday.  The inc/dec functions that take a value aren't just "add/sub one" they're "add/sub one and wrap at `val`", they have a very different use and behavior than `atomicAdd/Sub`.  The cuda and hip versions do that by default, the generic version is incomplete, and this one should either be fixed to do what the cuda and hip versions do or be removed to be fixed when the generic version is fixed.

Reply:
Forgot to cross-link the above to the RAJA issue where we ran into this, we expose that in RAJA for things like circular buffer management and possibly-overflowing signed counters.

llnl/raja#1073

Reply:
Oh, you're right. I totally misread the documentation.

Reply:
Since there are no builtins for `SYCL` for this operation, let's just fall back to the generic version and delete it here.

Reply:
Did you have to add it to get the code to compile?  If not just drop it and we can deal with CUDA and HIP elsewhere.

Reply:
Works for me, would you create an issue to check if one of the cl/sycl builtins in llvm might be usable for the purpose though? I can't recall and don't want to forget to at least give it a chance at better performance.

Reply:
See https://github.com/desul/desul/issues/29.